util/testutil/integration/dockerd.go

```diff
@@ -0,0 +1,107 @@
+package integration
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/docker/docker/client"
+)
+
+const dockerdBinary = "dockerd"
+
+// InitDockerdWorker registers a dockerd worker with the global registry.
+func InitDockerdWorker() {
+	Register(&dockerd{})
+}
+
+type dockerd struct{}
+
+func (c dockerd) Name() string {
+	return dockerdBinary
+}
+
+func (c dockerd) New(cfg *BackendConfig) (b Backend, cl func() error, err error) {
+	if err := lookupBinary(dockerdBinary); err != nil {
+		return nil, nil, err
+	}
+	if err := lookupBinary("buildkitd"); err != nil {
+		return nil, nil, err
+	}
+	if err := requireRoot(); err != nil {
+		return nil, nil, err
+	}
+
+	deferF := &multiCloser{}
+	cl = deferF.F()
+
+	defer func() {
+		if err != nil {
+			deferF.F()()
+			cl = nil
+		}
+	}()
+
+	tmpdir, err := ioutil.TempDir("", "bktest_dockerd")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	deferF.append(func() error { return os.RemoveAll(tmpdir) })
+
+	configFile := filepath.Join(tmpdir, "config.toml")
+	address := filepath.Join(tmpdir, "dockerd.sock")
+	cfgFile, err := os.Create(configFile)
+	if err != nil {
+		return nil, nil, err
+	}
+	err = json.NewEncoder(cfgFile).Encode(struct {
+		Debug  bool     `json:"debug"`
+		Socket string   `json:"socket"`
+		Hosts  []string `json:"hosts"`
+	}{
+		Debug:  true,
+		Socket: address,
+		Hosts:  []string{address},
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cmd := exec.Command(dockerdBinary, "--config", configFile)
```

Comment:
If we are still launching `dockerd` binary manually then I don't know the reason for updating the moby vendor and moving the test helpers out of internal there(possibly it is just for the `DialHijack` method). I'm fine with doing this directly in here but it would need much more logic from what `NewDaemon` does in moby. Eg. they need to use different storage, exec-root etc. Also `dockerd` has no `--config` flag afaik.

Reply:
Ah yes, flag appears to be `--config-file`, thanks. Looks like I just typoed that after looking it up. I thought we couldn't use the dockerd client package after all because it required a `testing.T` and this had to be used outside of tests, but now that it's only being used in tests I could probably refactor to pass in the `testing.T`. I'll look into it.

Reply:
Don't refactor these utility functions to take `testing.T`. Especially the methods in the `Sandbox/Worker` interfaces. If needed, refactor the `NewDaemon` helpers to have non-testing variants.

Reply:
I started working on refactoring daemon to not require a `testingT` and it's not a problem, it can be finished, but this does drastically increase the size of the change. Making this dependency between something that was designed for testing and code that isn't part of testing, is a risk in my mind and I'm wondering if we should reevaluate how this is done. I know it will make things take a lot longer, but would it be worth having a docker daemon package that's not tied to testing and refactoring the testing package to use that but log errors?

Reply:
Oh crap, there already is a `docker/daemon` package that I either already ruled out for some reason or missed entirely. Looking into how that works now.

Reply:
> this could also be use dot make the changes more incremental.

I didn't get exactly what you had in mind here. As `moby/testutil/daemon` doesn't actually seem to use `testing.T` for anything critical, easiest would be probably to just define new public functions that don't take testing but leave the old ones(that would now just call the generic ones) there so we don't have to refactor the tests atm.

> Oh crap, there already is a docker/daemon package that I either already ruled out for some reason or missed entirel

What pkg are you looking at? https://github.com/moby/moby/tree/master/daemon is implementation for some of the docker API so doesn't deal with `dockerd` binaries directly.

Reply:
> What pkg are you looking at? https://github.com/moby/moby/tree/master/daemon is implementation for some of the docker API so doesn't deal with dockerd binaries directly.

Yah, just went and looked and realized it wasn't the same thing at all. The naming confused me.

> I didn't get exactly what you had in mind here. As moby/testutil/daemon doesn't actually seem to use testing.T for anything critical, easiest would be probably to just define new public functions that don't take testing but leave the old ones(that would now just call the generic ones) there so we don't have to refactor the tests atm.

I could do this, but it feels pretty jank. We'd be doubling the API surface for no real benefit. I know moby doesn't provide any API guarantees, but in some ways this will make it even harder to get started learning the project because there's so much stuff to sort through. To a certain extent when we don't provide API stability guarantees we're inviting large changes because any small API change can lead to having to change everywhere it's used, but I think we should avoid making this problem worse by adding new APIs unnecessarily (which is more stuff that might need to be changed later). This probably would be the quickest way to get this particular change done though, so I'm not sure what's best.

Reply:
Went ahead and did this in https://github.com/moby/moby/pull/39943

Reply:
This will be a bit hard to test until the docker changes are merged (vendor appears to have issues so I can't build with -mod=vendor and docker doesn't have a go.mod file so the replace directive doesn't work, even if I add one logrus causes problems). I'll dig in and try to test it soon, but for now marking it as resolved to clear up these conversations.